File: service2.cpp
#include <iostream>
#include <sstream>
#include <fstream>
#include <thread>
#include <mutex>
#include <cassert>
#include <algorithm>

#include <sys/stat.h>

enum ERROR { NOT_FIFO=1, NO_CONTEXT };

class Service
{
protected:
    bool _has_current_context;
    std::mutex _mutex;
    std::string _file_current_context;
    std::string _file_data;
    std::string _out;
    std::string _current_context;

    bool has_current_context()
    {
        std::lock_guard<std::mutex> guarded_scope(_mutex);
        return _has_current_context;
    }

    void has_current_context(bool flag)
    {
        std::lock_guard<std::mutex> guarded_scope(_mutex);
        _has_current_context = flag;
    }

public:

    Service(
        std::string context,
        std::string data,
        std::string out
    ) :
        _has_current_context(false),
        _file_current_context(context),
        _file_data(data),
        _out(out)
    {}

    std::string strip(std::string s)
    {
        s.erase(std::find_if( s.rbegin(), s.rend(),
                    [](int ch) { return !std::isspace(ch); }
                    ).base(), s.end());
        return s;
    }

    void update_current_context()
    {
        while(true) {
            std::clog << "Wait for context..." << std::endl;
            bool has_error = false;
            try {
                std::ifstream if_current_context(_file_current_context);
                std::stringstream current_context;
                current_context << if_current_context.rdbuf();
                if_current_context.close();
                _current_context = strip(current_context.str());
            } catch(...) {
                has_error = true;
            }
            if(not has_error) {
                this->has_current_context(true);
                std::clog << "\tReceived context: " << _current_context << std::endl;
            }
        }
    }

    void handle_data()
    {
        while(true) {
            if(this->has_current_context()) {
                std::string data;
                std::clog << "Wait for data..." << std::endl;
                bool has_error = false;
                try {
                    std::ifstream if_data(_file_data);
                    std::stringstream datas;
                    datas << if_data.rdbuf();
                    if_data.close();
                    data = strip(datas.str());
                } catch(...) {
                    has_error = true;
                }
                if(not has_error) {
                    std::clog << "\tReceived data: " << data << std::endl;

                    std::clog << "Do stuff..." << std::endl;
                    std::ostringstream result;
                    result << _current_context << ": " << data;
                    std::clog << "\tdone" << std::endl;

                    std::clog << "Output..." << std::endl;
                    std::ofstream out(_out);
                    out << result.str() << std::endl;
                    out.close();
                    std::clog << "\tdone" << std::endl;
                } // if not has_error
            }
        } // while true
    }
};

bool is_named_pipe_fifo(char* filename)
{
    struct stat st;
    stat(filename, &st);
    if(not S_ISFIFO(st.st_mode) ) {
        return false;
    }
    return true;
}

int main(int argc, char** argv)
{
    assert(argc = 3);

    for(size_t i=1; i < 3; ++i) {
        if( not is_named_pipe_fifo(argv[i]) ) {
            std::cerr << "ERROR: " << argv[i] << " is not a named pipe FIFO" << std::endl;
            exit(ERROR::NOT_FIFO);
        }
    }

    std::clog << "Start server" << std::endl;
    std::clog.flush();
    Service server(argv[1], argv[2], argv[3]);

    std::thread do_current_context(&Service::update_current_context, &server);
    std::thread do_tasks(&Service::handle_data, &server);

    do_current_context.join();
    do_tasks.join();
    std::clog << "End" << std::endl;

}

File: service2.py
#!/usr/bin/env python

from enum import Enum
import threading
import stat
import sys
import os

class ERROR(Enum):
    NOT_FIFO = 1

class Service:
    def __init__(self, context: str, data: str, out: str) -> None:
        self._has_current_context: bool = False
        self._mutex = threading.Lock()
        self._file_current_context: str = context
        self._file_data: str = data
        self._out: str = out
        self._current_context: str = ""

    def get_has_current_context(self) -> bool:
        return self._has_current_context

    def set_has_current_context(self, flag: bool) -> None:
        self._mutex.acquire()
        self._has_current_context = flag
        self._mutex.release()

    def update_current_context(self) -> None:
        while True:
            print("Wait for context...", file = sys.stderr)
            has_error: bool = False
            try:
                with open(self._file_current_context) as if_current_context:
                    self._current_context: str = if_current_context.readline().strip()
            except:
                has_error = True

            if not has_error:
                self.set_has_current_context(True)
                print("\tReceived context:", self._current_context, file=sys.stderr)

    def handle_data(self) -> None:
        while True:
            if self.get_has_current_context():
                print("Wait for data...", file=sys.stderr)
                has_error: bool = False
                try:
                    with open(self._file_data) as if_data:
                        data: str = if_data.readline().strip()
                except:
                    has_error = True

                if not has_error:
                    print("\tReceived data:",data, file=sys.stderr)

                    print("Do stuff...", file=sys.stderr)
                    result: str = self._current_context + ":" + data
                    print("\tdone", file=sys.stderr)

                    print("Output...", file=sys.stderr)
                    with open(self._out, 'w') as out:
                        out.write(result)

                    print("\tdone", file=sys.stderr)

def is_named_pipe_fifo(filename: str):
    st = os.stat(filename)
    return stat.S_ISFIFO(st.st_mode)

if __name__ == "__main__":

    assert(len(sys.argv) == 4)

    for i in range(1,4):
        if not is_named_pipe_fifo(sys.argv[i]):
            print("ERROR:", sys.argv[i], "is not a named pipe FIFO", file=sys.stderr)
            sys.exit(ERROR.NO_FIFO)

    print("Start server", file=sys.stderr, flush=True)
    server = Service(sys.argv[1], sys.argv[2], sys.argv[3])

    do_current_context = threading.Thread( target = server.update_current_context )
    do_tasks           = threading.Thread( target = server.handle_data )

    do_current_context.start()
    do_tasks.start()

    do_current_context.join()
    do_tasks.join()

    print("End", file=sys.stderr)

